---
title: "Genotyping Report"
author: ""
output:
  html_document:
    toc: yes
    number_sections: no
    toc_float: no
    code_folding: hide
editor_options:
  markdown:
    wrap: 72
  chunk_output_type: inline
params:
  RData_image: ""
---
  
```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, include = TRUE)
```

```{r 'Molecular surveillance of drug resistance', include = FALSE}
library(dplyr)
library(magrittr)
load(params$RData_image)
```

## Performance Summary

This report summarizes data processed with the 01_Illumina_Amplicon_Processing 
workflow. It details the performance of amplicons and samples and the occurrence of 
mutations associated with resistance to different antimalarial drugs.

```{r, fig.width=7, fig.height=5, fig.cap = "**Figure 1:** Number of samples per location that yielded usable data."}
samples_by_location
```

```{r, fig.width=7, fig.height=5, fig.cap = "**Figure 2:** Dataset completeness. Sample success rate based on the provided thresholds of minimum read depth coverage per allele (ASV or amplicon sequence variant observed in the sample). The x-axis shows the percentage of amplicons exhibiting at least one ASV with read coverage at or above the minimum threshold. The y-axis shows the percentage of samples that amplify a specific percentage of loci given the ASV coverage threshold. The plot shows all analyzed samples."}
plot_precentage_of_samples_over_min_abd_reduced
```

```{r}
print(paste0("Filter ", 100*sample_ampl_rate, "% of amplified loci with read depth ", min_abd))
print(paste0(amplified_samples, "% of samples pass the amplification filter."))
```

## Genotype Frequency

```{r, fig.cap = "**Table 1:** Summary of resistance markers and their prevalence among samples that yielded usable data. The Annotation column specifies the drug for which the associated marker confers resistance. The Gene column indicates the gene in which the marker is identified. Count indicates the number of times the resistance markers were detected among the sample that yielded usable data. Samples_Genotyped provides the total number of samples in which the gene was successfully amplified, serving as the denominator for calculating marker prevalence. For some genes, resistance may be the results of mutations in several positions, which may not be validated."}
minimal_aa_table %>%
  DT::datatable(extensions = 'Buttons',
                options = list(dom = 'Blfrtip',
                               buttons = c('csv', 'excel')))
```

# Prevalence of resistant marker

```{r, fig.width=10, fig.height=10} 
if(!is.null(Longitude) & !is.null(Latitude)){
  library(tmap)
  tmap_mode('view')
  drug_resistant_haplotypes_plot$i_drug_map  
}
```
